### ISSUE TITLE
Stage level resource management to handle offheap/onheap memory conflict

### ISSUE THREAD

**Issue by FelixYBW** [oap-project/gluten#4392]
### Description

As Gluten currently can't fully cover spark functions, we will have to fallback some operators to Vanilla Spark, which leads an conflict of offheap and onheap memory conflict. Currently the suggested solution is that to those queries have both fallback and native operators, we need to config a large offheap memory and a large onheap memory, then the executor number should be decreased due to memory size constraint. 

The other solution is to spill offheap memory to onheap memory which we have implemented but performance isn't good, it's not used.

The 3rd solution is to use Spark3.0's stage level resource management, when we detect a stage has fallback, we can fallback the full stage, then restart the executors with high onheap memory.

@zhouyuan @PHILO-HE

**Comment by PHILO-HE** [comment]
For the 3rd solution, it depends on the stability of our whole stage fallback policy. I will investigate the feasibility. Thanks!

**Comment by FelixYBW** [comment]
@PHILO-HE  have you created a draft PR for this? any insight why the stage level resource manager can't be used by sparkSQL directly?

**Comment by SinghAsDev** [comment]
Also curious if this can be done in Apache Spark itself, there is some relevant discussion in this [SPIP](https://docs.google.com/document/d/1hxEPUirf3eYwNfMOmUHpuI5dIt_HJErCdo7_yr9htQc/edit?pli=1#heading=h.b6le5qnpmuht).

**Comment by FelixYBW** [comment]
@zhli1142015 Here is the related talk. Could you @ the guy so we can sync all here?

**Comment by FelixYBW** [comment]
Maybe it's known feature. We encountered a case in Spark3.5. Once you enable offheap in Spark, Spark will try to use offheap to cache the internalrows, if offheap memory isn't enough, it won't check the onheap memory but just spill the data. So if you config a large on heap memory but small offheap memory, spill will be triggered more frequently.

While you either can't config a large offheap memory and a small onheap memory because not all operators are able to cache the internal rows into offheap. Onheap demand is still very large for stages fallback to vanilla spark.

**Comment by philo-he** [comment]
> Maybe it's known feature. We encountered a case in Spark3.5. Once you enable offheap in Spark, Spark will try to use offheap to cache the internalrows, if offheap memory isn't enough, it won't check the onheap memory but just spill the data. So if you config a large on heap memory but small offheap memory, spill will be triggered more frequently.
> 
> While you either can't config a large offheap memory and a small onheap memory because not all operators are able to cache the internal rows into offheap. Onheap demand is still very large for stages fallback to vanilla spark.

@FelixYBW, we didn't realize this issue before. In Gluten's current resource profile rule, when whole stage fallback occurs, the offheap memory is reduced to 10% of the original value. Then, if I understand correctly, this setting can cause perf regression.

**Comment by zhztheplayer** [comment]
@PHILO-HE 

> when whole stage fallback occurs, the off-heap memory is reduced to 10% of the original value.

@FelixYBW 

> Onheap demand is still very large for stages fallback to vanilla spark.

Not sure if I understand correctly, but are you describing the same phenomenon? Where off-heap usage drops and on-heap usage rises, when the stage falls back?

IIRC Spark sort-based shuffle is a heavy operator that remains on on-heap when off-heap is on. I would be glad to do the some path-findings to see if we can somehow fix this. @FelixYBW If you want to also help locate where the remaining on-heap consumption came from? E.g., Could try setting `spark.shuffle.sort.bypassMergeThreshold = 2147483647` to disable vanilla sort-based shuffle then see if the on-heap consumption of vanilla Spark can be reduced. Thanks.

**Comment by FelixYBW** [comment]
> [@FelixYBW](https://github.com/FelixYBW), we didn't realize this issue before. In Gluten's current resource profile rule, when whole stage fallback occurs, the offheap memory is reduced to 10% of the original value. Then, if I understand correctly, this setting can cause perf regression.

You are right. with whole stage fallback, we may just config onheap memory and disable the offheap memory. 10% offheap memory definitely will lead to more spill.

We need a list of operators which still need large onheap memory when we enable offheap memory.

**Comment by FelixYBW** [comment]
> IIRC Spark sort-based shuffle is a heavy operator that remains on on-heap when off-heap is on. I would be glad to do the some path-findings to see if we can somehow fix this. [@FelixYBW](https://github.com/FelixYBW) If you want to also help locate where the remaining on-heap consumption came from? E.g., Could try setting `spark.shuffle.sort.bypassMergeThreshold = 2147483647` to disable vanilla sort-based shuffle then see if the on-heap consumption of vanilla Spark can be reduced. Thanks.

Not the shuffle. From the workload it's the sort operator. We enabled offheap but config a small offheap memory and large onheap memory, the spill is triggered even onheap memory is far enough for the sort.

**Comment by FelixYBW** [comment]
Since Spark has the unsaferow which can buffer data to offheap. I'd expect that all internalrow cacheing can be save to offheap. But the other struct has to use onheap. Like hashagg or hashjoin, the payload columns are able to save to offheap, but the hash table itself must be saved onheap. @JkSelf do you happen to know when you implement BHJ?

**Comment by JkSelf** [comment]
@FelixYBW Yes, for BHJ, we need to store the raw data for building the hash table in on-heap memory, and then have the driver distribute it to each executor. Even if we follow Spark's implementation approach of building the hash table only once on the driver side, we still need to store the hash table in on-heap memory before distributing it.

**Comment by FelixYBW** [comment]
> [@FelixYBW](https://github.com/FelixYBW) Yes, for BHJ, we need to store the raw data for building the hash table in on-heap memory, and then have the driver distribute it to each executor. Even if we follow Spark's implementation approach of building the hash table only once on the driver side, we still need to store the hash table in on-heap memory before distributing it.

Do you know for hash join, if worker build the hash table in onheap memory or is able to build in offheap?

**Comment by JkSelf** [comment]
@FelixYBW Yes, Gluten has implemented [UnsafeColumnarBuildSideRelation](https://github.com/apache/incubator-gluten/blob/ac9833fbc9e532e0435d91c41991cccf3f2ea742/backends-velox/src/main/scala/org/apache/spark/sql/execution/unsafe/UnsafeColumnarBuildSideRelation.scala#L88-L97) which enables building hash tables in off-heap memory. Note that this feature is currently experimental and the default setting is disabled (false).

**Comment by yantzu** [comment]
> The other solution is to spill offheap memory to onheap memory which we have implemented but performance isn't good, it's not used.

@FelixYBW Hi, regarding to this one, is there any report we can reference?

**Comment by FelixYBW** [comment]
> [@FelixYBW](https://github.com/FelixYBW) Hi, regarding to this one, is there any report we can reference?

https://github.com/apache/incubator-gluten/blob/main/docs/get-started/VeloxStageResourceAdj.md

**Comment by FelixYBW** [comment]
Meanwhile we are evaluating if it's possible to set onheap:offheap=1:3 for all the workload, which needs spark to allocate the large block memory in offheap always. @zhztheplayer already noted some allocations are fixing. Welcome to share your real-workload cases.
